core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -3285,13 +3287,31 @@ public void run(Reactor session) throws Exception {
      * Save the settings to a file.
      */
     public synchronized void save() throws IOException {
-        if(BulkChange.contains(this))   return;
-
-        if (initLevel == InitMilestone.COMPLETED) {
+        InitMilestone currentMilestone = initLevel;
+
+        if (!configLoaded) {
+            // someone is trying to save the config before all extensions are loaded (and possibly after as the task
+            // may run in parallel with other tasks.  OMG...!!! this is generally very bad and can lead to dataloss
+            LOGGER.log(Level.SEVERE,
+                       "An attempt to save Jenkins'' global configuration before it has been loaded has been "
+                       + "made during milestone " + currentMilestone
+                       + ".  This is indicative of a bug in the caller and may lead to full or partial loss of "
+                       + "configuration.",
+                       new IllegalStateException("call trace"));
+            // at this point we may want to terminate but the save may be called from a different thread and we
+            // can not call System.halt() because we could be running in a container :(
+            // for now just deny the save (the data will be replaced when we do load anyway
+            throw new IOException("An attempt to save the configuration was made before it was loaded");
```

Comment:
I'm expecting that this will break some unit tests that mock the lifecycle, using CI to find those.

Reply:
it did and they where fixed in 35552ca2e2a63a486fed9c5a5c21335e314655c8